keps/sig-storage/1790-recover-resize-failure/README.md

```diff
@@ -103,75 +112,99 @@ reduces the PVC request size, for both CSI and in-tree plugins they are designed
-- Expansion controller sees that CSI driver does not have `GET_VOLUME` controller capability.
-- `pvc.Status.AllocatedResources` still reports `100Gi` and hence there is no change in used quota.
+- Expansion controller starts expanding the volume and sees that last expansion failed on the node and driver does not have control-plane expansion.
+- Expansion controller sets `pvc.Status.AllocatedResources` to `20G`.
+- Expansion controller also sets `pvc.Status.ResizeStatus` to `ControllerExpansionInProgress`.
+- Since expansion operations in control-plane are NO-OP, expansion in control-plane succeeds and `pv.Spec` is set to `20G`.
+- Expansion succeed on the node with latest `allocatedResources` and `pvc.Status.Size` is set to `20G`.
+- Expansion controller also sets `pvc.Status.ResizeStatus` to `NodeExpansionPending`.
+- Kubelet can now retry expansion and expansion on node succeeds.
+- Kubelet sets `pvc.Status.ResizeStatus` to empty string and `pvc.Status.Capacity` to new value.
+- Quota controller sees a reduction in used quota because `max(pvc.Spec.Resources, pvc.Status.AllocatedResources)` is 20Gi.
 
 
 ##### Case 3 (Malicious user)
 - User increases 10Gi PVC to 100Gi by changing `pvc.spec.resources.requests["storage"] = "100Gi"`
-- `pvc.Status.AllocatedResources` still reports `10Gi`.
 - Quota controller uses `max(pvc.Status.AllocatedResources, pvc.Spec.Resources)` and adds `90Gi` to used quota.
 - Expansion controller slowly starts expanding the volume and sets `pvc.Status.AllocatedResources` to `100Gi` (before expanding).
+- Expansion controller also sets `pvc.Status.ResizeStatus` to `ControllerExpansionInProgress`.
 - At this point -`pv.Spec.Capacity` and `pvc.Status.Capacity` stays at 10Gi until the resize is finished.
 - While the storage backend is re-sizing the volume, user requests size 20Gi by changing `pvc.spec.resources.requests["storage"] = "20Gi"`
-- Expansion succeeds previous expansion and `pvc.Status.Capacity` and `pv.Spec.Capacity` report new size as `100Gi`.
-- Expansion controller notices that `pvc.Spec.Resources` < `pvc.Status.AllocatedResources` (meaning user tried to reduce size).
-- Expansion controller fetches actual size of the volume using `ControllerGetVolume` CSI RPC call.
-- Expansion controller sees that `actual_volume_size`(100Gi) >= `pvc.Status.AllocatedResources` (100Gi), so it does not make any modifictions to `pvc.Status.AllocatedResources`.
+- Expansion controller notices that previous expansion to last known `allocatedresources` is still in-progress.
+- Expansion controller keeps expanding the volume to `allocatedResources`.
+- Expansion to `100G` succeeds and `pv.Spec.Capacity` and `pvc.Status.Capacity` report new size as `100G`.
+- Although `pvc.Spec.Resources` reports size as `20G`, expansion to `20G` is never attempted.
+- Quota controller sees no change in storage usage by the PVC because `pvc.Status.AllocatedResources` is 100Gi.
+
+##### Case 4(Malicious User and rounding to GB/GiB bounaries)
```

Comment:
Do we need the new driver call to return actual size anymore? What if user increases size quickly twice in a row, and then lowers back down (eg to 50GB, 100GB, back down to 51GB)? Can we detect if the first expansion success is for the 50GB or 100GB?

I think since we don't set allocatedResources until right before we start expanding should be able to handle this?

Reply:
Yes in general as such we don't need drivers to report *actual* size in `ControllerExpandVolume` RPC call's response, but  `pv.Spec.Capacity` is set to value reported by the driver - so if they do start reporting it, it will be used.

And also yeah - I think rapidly expanding and then shrinking works fine because we don't set `allocatedResources` until we start expanding the volume.

Reply:
Can you add a flow for this rapid expansion then shink case? ie something like:
1. user increases size to 50GB
2. resize to 50GB starts but still in progress
3. user increases size to 100GB
4. first resize operation timesout for some reason
5. ...

Reply:
added